sda-commons-server-kafka/src/test/java/org/sdase/commons/server/kafka/consumer/strategies/MessageListenerStrategyOnCloseTest.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.sdase.commons.server.kafka.consumer.IgnoreAndProceedErrorHandler;
+import org.sdase.commons.server.kafka.consumer.MessageHandler;
+import org.sdase.commons.server.kafka.consumer.strategies.autocommit.AutocommitMLS;
+import org.sdase.commons.server.kafka.consumer.strategies.retryprocessingerror.RetryProcessingErrorMLS;
+import org.sdase.commons.server.kafka.consumer.strategies.synccommit.SyncCommitMLS;
+import org.sdase.commons.server.kafka.prometheus.ConsumerTopicMessageHistogram;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class MessageListenerStrategyOnCloseTest {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(MessageListenerStrategyOnCloseTest.class);
+  private static final String TOPIC_NAME = "topic";
+
+  @Test
+  void shouldCommitLastOffsetOnCloseForAutocommitMLS() {
+    final MyMessageHandler messageHandler = new MyMessageHandler();
+    shouldCommitLastOffsetOnClose(
+        new AutocommitMLS<String, String>(messageHandler, new IgnoreAndProceedErrorHandler<>()),
+        messageHandler);
+  }
+
+  @Test
+  void shouldCommitLastOffsetOnCloseForSyncCommitMLS() {
+    final MyMessageHandler messageHandler = new MyMessageHandler();
+    shouldCommitLastOffsetOnClose(
+        new SyncCommitMLS<String, String>(messageHandler, new IgnoreAndProceedErrorHandler<>()),
+        messageHandler);
+  }
+
+  @Test
+  void shouldCommitLastOffsetOnCloseForRetryProcessingErrorMLS() {
+    final MyMessageHandler messageHandler = new MyMessageHandler();
+    shouldCommitLastOffsetOnClose(
+        new RetryProcessingErrorMLS<String, String>(
+            messageHandler, new IgnoreAndProceedErrorHandler<>()),
+        messageHandler);
+  }
+
+  /**
+   * This is a test for the scenario when the application is shutdown while records are being
+   * processed. The strategy implementation must make sure that not the offset of the last poll is
+   * committed but the offset + 1 of the last record that was actually processed.
+   */
+  private void shouldCommitLastOffsetOnClose(
+      MessageListenerStrategy<String, String> strategy, MyMessageHandler messageHandler) {
+    // given
+    strategy.init(mock(ConsumerTopicMessageHistogram.class));
+
+    Map<TopicPartition, List<ConsumerRecord<String, String>>> recordsByPartition = new HashMap<>();
+    recordsByPartition.put(
+        new TopicPartition(TOPIC_NAME, 0),
+        Arrays.asList(
+            new ConsumerRecord<>(TOPIC_NAME, 0, 0, "key0", "value0"),
+            new ConsumerRecord<>(TOPIC_NAME, 0, 1, "key1", "value1"),
+            new ConsumerRecord<>(TOPIC_NAME, 0, 2, "key2", "value2")));
+
+    LOGGER.info("Starting thread for strategy");
+    final KafkaConsumer<String, String> consumer = mock(KafkaConsumer.class);
+    new Thread(
```

Comment:
Threads and timings always feel like they could introduce flakiness, but I assume this polling each 100ms and sleeping for 1s works out fine, right?

Reply:
Yeah, I also don't really like using "sleep" very much. But somehow I wanted to write a test that reproduces an application that does some longer processing and is shutdown in the middle of the process without being able to finish the other records from the last poll.